// js.html-builder/dist/HtmlBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var js_dsl_1 = require("js-dsl");
var lodash_1 = __importDefault(require("lodash"));
var tags = ['address', 'article', 'aside', 'footer',
    'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup', 'main', 'nav',
    'section', 'blockquote', 'dd', 'dir', 'div', 'dl', 'dt',
    'figcaption', 'figure', 'hr', 'li', 'ol', 'p', 'pre', 'ul', 'a', 'abbr',
    'b', 'bdi', 'bdo', 'br', 'cite', 'code',
    'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 'rb', 'rp', 'rt', 'rtc',
    'ruby', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup', 'time', 'tt',
    'u', 'var', 'wbr', 'area', 'audio', 'map', 'track', 'video',
    'applet', 'embed', 'iframe', 'noembed', 'object', 'param', 'picture',
    'source', 'canvas', 'noscript', 'script', 'del', 'ins', 'caption', 'col',
    'colgroup', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'button',
    'datalist', 'fieldset', 'form', 'input', 'label', 'legend', 'meter',
    'optgroup', 'option', 'output', 'progress', 'select', 'textarea', 'details',
    'dialog', 'menu', 'menuitem', 'summary', 'content', 'element', 'shadow',
    'slot', 'template', 'acronym', 'applet', 'basefont', 'bgsound', 'big',
    'blink', 'center', 'command', 'content', 'dir', 'element', 'font', 'frame',
    'frameset', 'image', 'isindex', 'keygen', 'listing', 'marquee', 'menuitem',
    'multicol', 'nextid', 'nobr', 'noframes', 'plaintext', 'shadow',
    'spacer', 'strike', 'tt', 'xmp', 'area', 'audio', 'map', 'track',
    'video', 'text'];
/**
 * Represents an element node in the DSL's object hierarchy.
 */
var Element = /** @class */ (function () {
    function Element(opts, attributes, value) {
        /**
         * This is a void element, no closing tag is required
         * @type {boolean}
         */
        this.noClosingTag = false;
        this.attributes = {};
        /**
         * Child elements
         */
        this.children = [];
        this.isComment = false;
        this.tag = opts.tag;
        this.noClosingTag = opts.noClosingTag || false;
        this.isComment = opts.isComment || false;
        if (typeof attributes != "object") {
            value = attributes;
        }
        else {
            this.attributes = attributes;
        }
        this.value = value;
    }
    Element.prototype.addChild = function (child) {
        this.children.push(child);
    };
    Element.prototype.toHtml = function (prefix) {
        if (prefix === void 0) { prefix = ""; }
        var res = [];
        if (this.tag === 'html') {
            res.push("<!DOCTYPE html>\n");
        }
        else if (this.tag !== 'text') {
            res.push('\n', prefix);
        }
        if (this.isComment) {
            res.push('<!--');
        }
        else if (this.tag !== 'text') {
            res.push("<" + this.tag);
            lodash_1.default.forOwn(this.attributes, function (v, k) { return res.push(" " + k + "=\"" + v + "\""); });
            res.push('>');
        }
        if (this.value) {
            res.push(String(this.value));
        }
        this.children.forEach(function (c) {
            res.push(c.toHtml(prefix + '  '));
        });
        if (!this.noClosingTag) {
            if (this.children.length && this.tag !== 'text') {
                res.push('\n');
                res.push(prefix);
            }
            if (this.isComment) {
                res.push("-->");
            }
            else if (this.tag !== 'text') {
                res.push("</" + this.tag + ">");
            }
        }
        return res.join('');
    };
    return Element;
}());
exports.Element = Element;
var ElementFactory = /** @class */ (function (_super) {
    __extends(ElementFactory, _super);
    function ElementFactory(opts) {
        if (opts === void 0) { opts = {}; }
        var _this = _super.call(this) || this;
        _this.opts = opts;
        return _this;
    }
    // noinspection JSUnusedGlobalSymbols
    ElementFactory.prototype.newInstance = function (builder, name, attr, value) {
        var opts = {
            tag: name,
            noClosingTag: this.opts.noClosingTag,
            isComment: this.opts.isComment
        };
        return new Element(opts, attr, value);
    };
    // noinspection JSMethodCanBeStatic,JSUnusedGlobalSymbols
    ElementFactory.prototype.setChild = function (builder, parent, child) {
        parent.addChild(child);
    };
    return ElementFactory;
}(js_dsl_1.AbstractFactory));
exports.ElementFactory = ElementFactory;
var HtmlFactory = /** @class */ (function (_super) {
    __extends(HtmlFactory, _super);
    function HtmlFactory() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSMethodCanBeStatic,JSUnusedGlobalSymbols
    HtmlFactory.prototype.getBuilder = function (parent) {
        return new MainBuilder(parent);
    };
    return HtmlFactory;
}(ElementFactory));
var HeadFactory = /** @class */ (function (_super) {
    __extends(HeadFactory, _super);
    function HeadFactory() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSMethodCanBeStatic,JSUnusedGlobalSymbols
    HeadFactory.prototype.getBuilder = function (parent) {
        return new HeadBuilder(parent);
    };
    return HeadFactory;
}(ElementFactory));
var BlockFactory = /** @class */ (function (_super) {
    __extends(BlockFactory, _super);
    function BlockFactory() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSMethodCanBeStatic,JSUnusedGlobalSymbols
    BlockFactory.prototype.getBuilder = function (parent) {
        return new BlockBuilder(parent);
    };
    return BlockFactory;
}(ElementFactory));
exports.BlockFactory = BlockFactory;
var MainBuilder = /** @class */ (function (_super) {
    __extends(MainBuilder, _super);
    function MainBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSUnusedGlobalSymbols
    MainBuilder.prototype.register = function () {
        this.registerFactory('head', new HeadFactory());
        this.registerFactory('body', new BlockFactory());
    };
    return MainBuilder;
}(js_dsl_1.JsDsl));
exports.MainBuilder = MainBuilder;
var HeadBuilder = /** @class */ (function (_super) {
    __extends(HeadBuilder, _super);
    function HeadBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSUnusedGlobalSymbols
    HeadBuilder.prototype.register = function () {
        this.registerFactory('title', new ElementFactory());
        this.registerFactory('link', new ElementFactory({ noClosingTag: true }));
        this.registerFactory('meta', new ElementFactory({ noClosingTag: true }));
        this.registerFactory('style', new ElementFactory());
    };
    return HeadBuilder;
}(js_dsl_1.JsDsl));
exports.HeadBuilder = HeadBuilder;
var BlockBuilder = /** @class */ (function (_super) {
    __extends(BlockBuilder, _super);
    function BlockBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSUnusedGlobalSymbols
    BlockBuilder.prototype.register = function () {
        var _this = this;
        tags.forEach(function (b) { return _this.registerFactory(b, new ElementFactory()); });
        this.registerFactory('img', new ElementFactory({ noClosingTag: true }));
    };
    return BlockBuilder;
}(js_dsl_1.JsDsl));
exports.BlockBuilder = BlockBuilder;
var HtmlBuilder = /** @class */ (function (_super) {
    __extends(HtmlBuilder, _super);
    function HtmlBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // noinspection JSUnusedGlobalSymbols
    HtmlBuilder.prototype.register = function () {
        this.registerFactory('html', new HtmlFactory());
        this.registerFactory('comment', new ElementFactory({ isComment: true }));
        this.registerFactory('text', new ElementFactory());
    };
    return HtmlBuilder;
}(js_dsl_1.JsDsl));
exports.HtmlBuilder = HtmlBuilder;
/**
 * Accepts a configuration function which is responsible for specifying node
 * hierarchy. Builds the node hierarchy and returns the root element (html).
 *
 * @param {Config} f the configuration function
 * @returns {Element} the root element
 */
function htmlBuilder(f) {
    return new HtmlBuilder().build(f);
}
exports.htmlBuilder = htmlBuilder;
/**
 * Accepts a configuration function which is responsible for specifying node
 * hierarchy. Builds the node hierarchy and returns the root element (any block
 * element).
 *
 * @param {Config} f the configuration function
 * @returns {Element} the root element
 */
function snippetBuilder(f) {
    return new BlockBuilder().build(f);
}
exports.snippetBuilder = snippetBuilder;
